### DISCUSSION TITLE
Has anyone succesfully implemented a AuthProvider?

### DISCUSSION THREAD

**Discussion by OliverGeneser** [ory/.github#33] [Q&A]
Hello!
I have followed the NextJS guide and implemented a prototype of an AuthProvider, but I have a problem with rerenders and was wondering if someone has designed a working provider?

Something like this:
https://github.com/auth0/nextjs-auth0

Greetings
Oliver G.

**Comment by aeneasr** [comment]
We are working towards an SDK with React Hooks like this, but it's not yet available. Take a look at github.com/ory/elements though! :)

/cc @mszekiel @Benehiko

**Reply by OliverGeneser** [reply]
@aeneasr Nice! Is there any ETA? or roadmap to follow?

**Comment by hridaya2004** [comment]
bump.

@aeneasr is there any update on this? or am i supposed to use the ones in @ory/elements-react? aren't those just the components where we have to write the flows on our own?

**Comment by GitNimay** [comment]
I would avoid putting session fetching logic directly into a global React context provider if it causes the whole tree to update on every route change or session check.

In Next.js, a provider can work, but it should usually be very small and only wrap the client components that actually need client side session state. Server Components cannot consume React context anyway, so wrapping the entire app often gives you more rerenders without much benefit.

A pattern that works well is:

```tsx
"use client";

import {
  createContext,
  useContext,
  useMemo,
  useState,
  type ReactNode,
} from "react";

type AuthContextValue = {
  session: unknown | null;
  setSession: (session: unknown | null) => void;
};

const AuthContext = createContext<AuthContextValue | null>(null);

export function AuthProvider({
  children,
  initialSession,
}: {
  children: ReactNode;
  initialSession: unknown | null;
}) {
  const [session, setSession] = useState(initialSession);

  const value = useMemo(
    () => ({
      session,
      setSession,
    }),
    [session],
  );

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export function useAuth() {
  const value = useContext(AuthContext);

  if (!value) {
    throw new Error("useAuth must be used inside AuthProvider");
  }

  return value;
}
```

Then fetch the initial session on the server and pass it into the client provider:

```tsx
import { AuthProvider } from "@/components/auth-provider";
import { getSession } from "@/lib/auth";

export default async function Layout({
  children,
}: {
  children: React.ReactNode;
}) {
  const session = await getSession();

  return (
    <AuthProvider initialSession={session}>
      {children}
    </AuthProvider>
  );
}
```

The important part is that the provider should not call `setSession` during render, should not create a new context value object on every render without `useMemo`, and should not refetch the session from every consumer. Fetch once at the boundary, pass the stable initial value down, and expose an explicit `refreshSession()` only where needed.

For Ory specifically, I would usually structure it like this:

```text
Server side:
Read cookies
Call Kratos /sessions/whoami
Redirect if no valid session

Client side:
Use AuthProvider only for UI state that must update without navigation
For example user menu, logout button, or optimistic session refresh
```

So the equivalent of `@auth0/nextjs-auth0` would be less of a fully global client provider and more of a small wrapper around server side session checks plus a minimal client context.

If you share the provider code, it should be possible to point out the exact rerender trigger, but common causes are:

```text
1. Fetching the session inside the provider on every render
2. Passing a non memoized object as the context value
3. Putting router/pathname/search params in the provider state
4. Wrapping the whole app when only the navigation needs session state
5. Calling setState from effects whose dependencies change every render
```

I would start by moving session loading to the server layout or route handler, then keep the provider as a thin client side cache of that session.

**Reply by hridaya2004** [reply]
You can probably just use authts/react-oidc-context, and probably just add session within the context.

**Reply by GitNimay** [reply]
Sorry bro I m spamming the answer to get galaxy brain badge ....I don't really know how to fix that 🙏
